### ISSUE TITLE
Feature Request: Limited Visibility

### ISSUE THREAD

**Issue by zimt28** [absinthe-graphql/absinthe#710]
I'm suggesting a way to limit the visibility of parts of the schema. The GraphQL Ruby gem has added this feature so I'll quote:

> With GraphQL-Ruby, it’s possible to hide parts of your schema from some users. This isn’t exactly part of the GraphQL spec, but it’s roughly within the bounds of the spec.
>
> Here are some reasons you might want to hide parts of your schema:
>
> You don’t want non-admin users to know about administration functions of the schema.
You’re developing a new feature and want to make a gradual release to only a few users first.

https://graphql-ruby.org/authorization/visibility

**Comment by hisapy** [comment]
I think that maybe this kind of out-of-spec Feature should be implemented via _plugins_ or _middleware_ keeping the core modular and _relatively simple_ . I'm not very familiar with the current architecture of this lib though.

**Comment by benwilson512** [comment]
@hisapy right now it isn't possible to implement it yourself because absinthe special cases the introspection fields. In 1.5 the introspection fields will be added via a schema compilation phase, and then folks wishing to not have all of those fields are free to alter their schema's compilation phases to work otherwise.

**Comment by 1player** [comment]
On a project I'm working on I've per-field ACLs with a custom middleware. The idea is that everything is private-only, unless marked as public with a meta tag:

```
object :user do
    meta(public_fields: [:name])

    field(:name, :string)
    field(:email, :string)
  end
```

In this case unauthenticated users can fetch the `name` field, but `email` is only available if you're authenticated. Since the root query is an object, we limit each top-level query with the same logic.

This access control is done by a middleware (https://gist.github.com/1player/65576b65a046b1b43ac9705bacc51dab), which is added to the list of active middleware in the `middleware/3` callback:

```
def middleware(middleware, %{identifier: identifier} = field, object) do
    public_fields = Absinthe.Type.meta(object, :public_fields)

    middleware =
      [
        {Hive.API.Schema.Middleware.FieldACL, {identifier, public_fields}}
      ] ++ middleware

    middleware
  end
```

The general idea is everything we're resolving an object, we fetch its `public_field` meta, and pass it through the middleware. Depending on its value and whether there's a user logged in or not, we decide whether to continue resolution or fail with an error.

`public_fields` has been made a little more ergonomic by accepting the following values:

```
meta(public_fields: [:name]) # only name is public
meta(public_fields: :all) # all fields are public
meta(public_fields: {:except, [:name]}) # everything except name is public
```

That said, this is a bit of a hack, can definitely be improved, but is currently working well for our use case, and it's currently being used in production.

**Comment by benwilson512** [comment]
> had a hard time understanding whether it's possible to hook into the Absinthe internals to hide private fields.

It is not possible yet, but it will be!

Your public / provide ACL based field solution is very elegant, and I've actually suggested similar solutions to people before. Unfortunately though that doesn't solve the problem about introspection, just access.

**Comment by bruce** [comment]
Built-in limited visibility is not on the roadmap for the main Absinthe package (this repository). Please watch #711 for the introspection work @benwilson512 is referring to; once that is in:

- We would be interested in 3rd party packages that implement this behavior. While we intend on keeping Absinthe itself focused as tightly as possible on the official GraphQL specification behavior, we've taken (and will continue to take) significant efforts to make its internals as extensible as possible via our pipeline mechanism.
- We're happy to assist with questions on _how_ to extend things via Slack (or, eventually, we hope to move these types of discussions to the Elixir Forum for durability).
- If there's significant enough interest and the author is amenable, we're open to adopting packages officially.

**Comment by bruce** [comment]
(Actually, I'll reopen this and allow it to be closed by #711's merge)
